File: frontend/src/api/predictions.ts
import api from "./client";
import type {
  AllianceWeights,
  OptimalAlliancesResponse,
  ComplementResponse,
} from "../types";

export async function predictAlliances(
  eventKey: string,
  weights?: AllianceWeights
): Promise<OptimalAlliancesResponse> {
  const { data } = await api.post<OptimalAlliancesResponse>(
    "/api/predict/optimal-alliances",
    { event_key: eventKey, weights }
  );
  return data;
}

export async function fetchComplements(
  eventKey: string,
  teamKey: string
): Promise<ComplementResponse> {
  const { data } = await api.get<ComplementResponse>(
    `/api/complement/${eventKey}/${teamKey}`
  );
  return data;
}

File: backend/app/routers/predictions.py
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app.database import get_db
from app.models.team_event import TeamEvent
from app.schemas.prediction import (
    OptimalAlliancesRequest,
    OptimalAlliancesResponse,
)
from app.services.alliance_optimizer import AllianceOptimizer

router = APIRouter()


@router.post("/predict/optimal-alliances", response_model=OptimalAlliancesResponse)
async def predict_alliances(
    req: OptimalAlliancesRequest, db: Session = Depends(get_db)
):
    team_events = (
        db.query(TeamEvent)
        .filter(TeamEvent.event_key == req.event_key)
        .all()
    )

    optimizer = AllianceOptimizer(req.weights)
    alliances = optimizer.compute_optimal_alliances(team_events)

    return OptimalAlliancesResponse(
        event_key=req.event_key, alliances=alliances
    )
